File: packages/elixir/lib/kreuzberg/code_docstring_info.ex
defmodule Kreuzberg.CodeDocstringInfo do
  @moduledoc """
  Docstring information with parsed sections.

  ## Fields

    * `:text` - Raw docstring text
    * `:format` - Docstring format (e.g. "javadoc", "numpy", "google")
    * `:associated_item` - Name of the associated code item
    * `:span` - Source span
    * `:sections` - Parsed docstring sections
  """

  @type t :: %__MODULE__{
          text: String.t(),
          format: String.t(),
          associated_item: String.t() | nil,
          span: Kreuzberg.CodeSpan.t(),
          sections: [Kreuzberg.CodeDocSection.t()]
        }

  defstruct text: "",
            format: "",
            associated_item: nil,
            span: %Kreuzberg.CodeSpan{},
            sections: []

  @spec from_map(map()) :: t()
  def from_map(data) when is_map(data) do
    span =
      case Map.get(data, "span", Map.get(data, :span)) do
        %Kreuzberg.CodeSpan{} = s -> s
        map when is_map(map) -> Kreuzberg.CodeSpan.from_map(map)
        _ -> %Kreuzberg.CodeSpan{}
      end

    sections =
      case Map.get(data, "sections", Map.get(data, :sections, [])) do
        list when is_list(list) -> Enum.map(list, &Kreuzberg.CodeDocSection.from_map/1)
        _ -> []
      end

    %__MODULE__{
      text: Map.get(data, "text", Map.get(data, :text, "")),
      format: Map.get(data, "format", Map.get(data, :format, "")),
      associated_item: Map.get(data, "associated_item", Map.get(data, :associated_item)),
      span: span,
      sections: sections
    }
  end

  @spec to_map(t()) :: map()
  def to_map(%__MODULE__{} = info) do
    %{
      "text" => info.text,
      "format" => info.format,
      "associated_item" => info.associated_item,
      "span" => Kreuzberg.CodeSpan.to_map(info.span),
      "sections" => Enum.map(info.sections, &Kreuzberg.CodeDocSection.to_map/1)
    }
  end
end

File: packages/php/src/Types/CodeDocstringInfo.php
<?php

declare(strict_types=1);

namespace Kreuzberg\Types;

/**
 * Docstring information with parsed sections.
 *
 * @property-read string $text Raw docstring text
 * @property-read string $format Docstring format (e.g. 'javadoc', 'numpy', 'google')
 * @property-read string|null $associatedItem Name of the associated code item
 * @property-read CodeSpan $span Source span
 * @property-read array<CodeDocSection> $sections Parsed docstring sections
 */
readonly class CodeDocstringInfo
{
    /**
     * @param array<CodeDocSection> $sections
     */
    public function __construct(
        public string $text,
        public string $format,
        public ?string $associatedItem,
        public CodeSpan $span,
        public array $sections,
    ) {
    }

    /**
     * @param array<string, mixed> $data
     */
    public static function fromArray(array $data): self
    {
        /** @var string $text */
        $text = $data['text'] ?? '';

        /** @var string $format */
        $format = $data['format'] ?? '';

        /** @var string|null $associatedItem */
        $associatedItem = $data['associated_item'] ?? null;

        /** @var array<string, mixed> $spanData */
        $spanData = $data['span'] ?? [];

        /** @var array<array<string, mixed>> $sectionsData */
        $sectionsData = $data['sections'] ?? [];

        return new self(
            text: $text,
            format: $format,
            associatedItem: $associatedItem,
            span: CodeSpan::fromArray($spanData),
            sections: array_map(
                /** @param array<string, mixed> $section */
                static fn (array $section): CodeDocSection => CodeDocSection::fromArray($section),
                $sectionsData,
            ),
        );
    }
}
